### ISSUE TITLE
Socket does not exist when calling BasicPublish

### ISSUE THREAD

**Issue by markunsworth** [pika/pika#490]
Since updating to 0.9.14 we've seen in large number of these errors

    Traceback (most recent call last):
      File "/home/tweet-filter/events/controller.py", line 117, in process
        self.notification_events.send(msg)
      File "/home/tweet-filter/.env/local/lib/python2.7/site-packages/commons/rabbitmq/job_sender_sync.py", line 111, in send
        mandatory=True)
      File "/home/tweet-filter/.env/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 553, in basic_publish
        (properties, body))
      File "/home/tweet-filter/.env/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 1141, in _rpc
        self._wait_on_response(method_frame))
      File "/home/tweet-filter/.env/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 1159, in _send_method
        self.connection.send_method(self.channel_number, method_frame, content)
      File "/home/tweet-filter/.env/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 274, in send_method
        self._send_method(channel_number, method_frame, content)
      File "/home/tweet-filter/.env/local/lib/python2.7/site-packages/pika/connection.py", line 1503, in _send_method
        self._send_frame(frame.Method(channel_number, method_frame))
      File "/home/tweet-filter/.env/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 417, in _send_frame
        super(BlockingConnection, self)._send_frame(frame_value)
      File "/home/tweet-filter/.env/local/lib/python2.7/site-packages/pika/connection.py", line 1490, in _send_frame
        self._flush_outbound()
      File "/home/tweet-filter/.env/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 377, in _flush_outbound
        if self._handle_write():
      File "/home/tweet-filter/.env/local/lib/python2.7/site-packages/pika/adapters/base_connection.py", line 360, in _handle_write
        self.socket.sendall(frame)
    AttributeError: 'NoneType' object has no attribute 'sendall'

**Comment by DanCech** [comment]
I ran into this issue last week, and after looking around at the code I don't see any way to safely raise an exception from within `BlockingConnection._check_state_on_disconnect()`.

I experimented with using try/finally in `_adapter_disconnect()` to make sure `_init_connection_state()` was always called then bubble the exception up the stack, but that causes a problem in `_on_connection_closed()` because it means that the code to handle channel and callback teardown isn't executed.

In the end I'm using this workaround for now to catch the exception and things are working the way I expect.
```python
class BlockingConnection(pika.BlockingConnection):
    """
    This class works around a bug in pika.BlockingConnection's
    _check_state_on_disconnect() method; it raises an exception which prevents
    the connection state from being updated properly when the rabbit
    connection is dropped.  We catch and ignore the exception to allow
    _adapter_disconnect() to properly call _init_connection_state() and not
    leave the client in an inconsistent state.
    """
    def _check_state_on_disconnect(self):
        try:
            super(BlockingConnection, self)._check_state_on_disconnect()
        except pika.exceptions.ConnectionClosed:
            pass
```

**Comment by gmr** [comment]
Hopefully df7d3b742bb addresses what you all are running into here.

**Comment by lovato** [comment]
Hi @gmr 

I an running into this issue, and already saw it is supposed to be fixed under master branch.
But master differs from 0.9.14 a lot, which (I guess) makes a cherry pick near impossible. I am correct?

If so, is there any official patch to fix this for 0.9.14? Or can you recommend a specific cherry pick?

Best,
Lovato

**Comment by niteeshm** [comment]
@gmr ,

I am using a single pika connection and channel (global) and publishing the message inside a Celery Task and when I execute this task multiple times. It still gives me this error : 

[2015-07-03 16:02:37,315: ERROR/MainProcess] Task celery_rabbit.add[cea48b24-236c-4777-a62e-64854ed91737] raised unexpected: OSError(9, 'Bad file descriptor')
Traceback (most recent call last):
  File "/Library/Python/2.7/site-packages/celery/app/trace.py", line 240, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/Library/Python/2.7/site-packages/celery/app/trace.py", line 438, in __protected_call__
    return self.run(*args, **kwargs)
  File "/Users/8320/myntra/git/lgp-fumes/activity_engine/app/helpers/celery_rabbit.py", line 15, in add
    body='Hello World!')
  File "build/bdist.macosx-10.10-intel/egg/pika/adapters/blocking_connection.py", line 1916, in basic_publish
  File "build/bdist.macosx-10.10-intel/egg/pika/adapters/blocking_connection.py", line 2005, in publish
  File "build/bdist.macosx-10.10-intel/egg/pika/channel.py", line 338, in basic_publish
  File "build/bdist.macosx-10.10-intel/egg/pika/channel.py", line 1150, in _send_method
  File "build/bdist.macosx-10.10-intel/egg/pika/connection.py", line 1571, in _send_method
  File "build/bdist.macosx-10.10-intel/egg/pika/connection.py", line 1601, in _send_message
  File "build/bdist.macosx-10.10-intel/egg/pika/adapters/base_connection.py", line 282, in _flush_outbound
  File "build/bdist.macosx-10.10-intel/egg/pika/adapters/base_connection.py", line 452, in _handle_write
  File "build/bdist.macosx-10.10-intel/egg/pika/adapters/base_connection.py", line 338, in _handle_error
  File "build/bdist.macosx-10.10-intel/egg/pika/adapters/base_connection.py", line 288, in _handle_disconnect
  File "build/bdist.macosx-10.10-intel/egg/pika/adapters/select_connection.py", line 95, in _adapter_disconnect
  File "build/bdist.macosx-10.10-intel/egg/pika/adapters/select_connection.py", line 319, in remove_handler
  File "build/bdist.macosx-10.10-intel/egg/pika/adapters/select_connection.py", line 480, in update_handler
OSError: [Errno 9] Bad file descriptor

**Comment by vitaly-krugl** [comment]
@niteeshm, please enable pika DEBUG-level logging, reproduce this failure, and attach the full pika log to this issue.

**Comment by SMesser** [comment]
@niteeshm 's OSError appears to be a bit different from the AttributeError otherwise described on this report. I'm still seeing the original error, despite running pika 0.9.14. Here's my traceback:

<pre> File "/usr/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 687, in close
    self.basic_cancel(consumer_tag=consumer_tag)
  File "/usr/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 490, in basic_cancel
    self._on_cancelok, replies)
  File "/usr/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 1141, in _rpc
    self._wait_on_response(method_frame))
  File "/usr/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 1159, in _send_method
    self.connection.send_method(self.channel_number, method_frame, content)
  File "/usr/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 274, in send_method
    self._send_method(channel_number, method_frame, content)
  File "/usr/local/lib/python2.7/site-packages/pika/connection.py", line 1503, in _send_method
    self._send_frame(frame.Method(channel_number, method_frame))
  File "/usr/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 417, in _send_frame
    super(BlockingConnection, self)._send_frame(frame_value)
  File "/usr/local/lib/python2.7/site-packages/pika/connection.py", line 1490, in _send_frame
    self._flush_outbound()
  File "/usr/local/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 377, in _flush_outbound
    if self._handle_write():
  File "/usr/local/lib/python2.7/site-packages/pika/adapters/base_connection.py", line 360, in _handle_write
    self.socket.sendall(frame)
AttributeError: 'NoneType' object has no attribute 'sendall'
</pre>

**Comment by TeddyHartanto** [comment]
I had some tests that I want to run. Everything is fine when I run it in my localhost. But, when I run the tests in a server, I had exactly the same problem and situation as this one. I checked my pika version, it's 0.9.13. I'm using BlockingConnection too, by the way.

First, I tried to call process_data_events() periodically. It didn't work.
Second, I tried to open a new connection and channel whenever I publish a message or open a new queue. It didn't work.

@eandersson mentioned the possibility of a socket bug. So, I checked the python version, hypothesizing that maybe there IS a bug, and it was fixed in the later version of python. In the server, the python version is 2.7.3, and in my localhost it's 2.7.12. To test if python version is really the problem, I installed conda and created an environment with python version 2.7.3. I ran the tests and it passed (I can't reproduce the issue).

After all of the above attempts, I came up with another hypothesis that maybe it's a bug in my server's rabbitmq-server. I compared the versions: in localhost it's the latest (3.6.5), in my server it's 2.8.4. To verify that this is the actual problem, I ran the tests, but I used a remote rabbitmq with a later version. Everything worked fine. So, I upgraded the rabbitmq-server and, lo and behold, the problem disappears!

TL;DR:
The solution is to upgrade your rabbitmq-server!

**Comment by vitaly-krugl** [comment]
Pika 0.9.x is quite old. In particular, 0.10.0 had a major rewrite of the BlockingConnection.
